lib/ansible/plugins/connection/psrp.py

```diff
@@ -622,81 +551,47 @@ def fetch_file(self, in_path: str, out_path: str) -> None:
-        read_script = """param([int64]$Offset, [int]$BufferSize)
-$ErrorActionPreference = "Stop"
-$fs.Seek($Offset, [System.IO.SeekOrigin]::Begin) > $null
-$buffer = New-Object -TypeName byte[] -ArgumentList $BufferSize
-$read = $fs.Read($buffer, 0, $buffer.Length)
-
-if ($read -gt 0) {
-    [System.Convert]::ToBase64String($buffer, 0, $read)
-}"""
-
-        # need to run the setup script outside of the local scope so the
-        # file stream stays active between fetch operations
-        rc, stdout, stderr = self._exec_psrp_script(
-            setup_script,
-            use_local_scope=False,
-            arguments=[in_path],
-        )
-        if rc != 0:
-            raise AnsibleError("failed to setup file stream for fetch '%s': %s"
-                               % (out_path, to_native(stderr)))
-        elif stdout.strip() == '[DIR]':
-            # to be consistent with other connection plugins, we assume the caller has created the target dir
-            return
+        script, in_data = _bootstrap_powershell_script('psrp_fetch_file.ps1', {
+            'Path': in_path,
+            'BufferSize': buffer_size,
+        })
 
-        b_out_path = to_bytes(out_path, errors='surrogate_or_strict')
-        # to be consistent with other connection plugins, we assume the caller has created the target dir
-        offset = 0
-        with open(b_out_path, 'wb') as out_file:
-            while True:
-                display.vvvvv("PSRP FETCH %s to %s (offset=%d" %
-                              (in_path, out_path, offset), host=self._psrp_host)
-                rc, stdout, stderr = self._exec_psrp_script(
-                    read_script,
-                    arguments=[offset, buffer_size],
-                )
-                if rc != 0:
-                    raise AnsibleError("failed to transfer file to '%s': %s"
-                                       % (out_path, to_native(stderr)))
-
-                data = base64.b64decode(stdout.strip())
-                out_file.write(data)
-                if len(data) < buffer_size:
-                    break
-                offset += len(data)
-
-            rc, stdout, stderr = self._exec_psrp_script("$fs.Close()")
-            if rc != 0:
-                display.warning("failed to close remote file stream of file "
-                                "'%s': %s" % (in_path, to_native(stderr)))
+        ps = PowerShell(self.runspace)
+        ps.add_script(script)
+        ps.begin_invoke(in_data.decode().splitlines())
+
+        # Call poll once to get the first output telling us if it's a file/dir/failure
+        ps.poll_invoke()
+
+        if ps.output:
+            if ps.output[0] == '[DIR]':
+                # to be consistent with other connection plugins, we assume the caller has created the target dir
+                return
+
+            ps.output.pop(0)
+            with open(b_out_path, 'wb') as out_file:
+                while True:
```

Comment:
This feels dangerous unless `poll_invoke` has a really reliable timeout that doesn't default to infinite...

Reply:
It times out every WSMan OperationTimeout value (which I think we default to 30s). The trouble is how do we distinguish between a host just being slow vs something catastrophic happened on the other end.